### PR TITLE
Implement Pictures and Weather in Battle

### PR DESCRIPTION
**Pull request by fmatthew5876**
Depends on: #1989

- [x] Implement erase on battle end flag
- [x] Implement displaying pictures in battle
- [x] Implement displaying weather effects in battle
- [x] Rebase after #2004 merged
- [x] Rebase after #2012 merged and fix `Game_Screen` for #1954 

Fix: #1488 
Fix: #1263
Progress on: #1954

### CODE REVIEW

**Comment by fmatthew5876** [comment]
This is easy to test, create an event which makes these pictures:

| PicId | OnMap | OnBattle | EraseOnMapChange | EraseOnBattleEnd |
| -- | -- | -- | -- | -- |
| 1 | 1 | 0 | 0 | 0 |
| 2 | 1 | 0 | 1 | 0 |
| 3 | 1 | 0 | 0 | 1 |
| 4 | 1 | 0 | 1 | 1 |
| | | | |
| 5 | 0 | 1 | 0 | 0 |
| 6 | 0 | 1 | 1 | 0 |
| 7 | 0 | 1 | 0 | 1 |
| 8 | 0 | 1 | 1 | 1 |
| | | | |
| 9 | 1 | 1 | 0 | 0 |
| 10 | 1 | 1 | 1 | 0 |
| 11 | 1 | 1 | 0 | 1 |
| 12 | 1 | 1 | 1 | 1 |

1. Run the event on a map, start and finish a battle, the start another battle

Result: 
* Pics 1-4, 9-12 visible on map
* pics 5-12 visible in battle
* pics 1,2,9,10 visible on map after
* pics 5,6,9,10 visible in next battle

2. Run the event in a battle, finish the battle, start a new battle

Result:
* pics 5-12 visible in battle
* pics 1,2,9,10 visible on map after
* pics 5,6,9,10 visible in next battle

3. Run the event on a map, and teleport to the same map
* Pics 1-4, 9-12 visible on map
* Pics 1-4, 9-12 still visible after teleport

4. Run the event on a map, and teleport to a different map
* Pics 1-4, 9-12 visible on map
* Pics 1,3,9,11 visible on other map

**Comment by fmatthew5876** [comment]
Would prefer #2004 goes in before this one

**Comment by fmatthew5876** [comment]
Rebased and now was finally able to add the commit removing `Main_Data::game_data.screen`.

I did redid the battle and map change picture tests in my test game. I also tested saving and loading screen state.

This is ready once #1989 goes in.

File: src/game_picture.h [hunk 1]
```diff
@@ -69,15 +70,29 @@ class Game_Picture {
...
 	static void UpdateSprite(std::vector<Game_Picture>& pictures);
 
-	void Update();
-	void UpdateSprite();
+	void OnMapChange();
+	void OnBattleStart(Scene* map_scene, Scene& battle_scene);
+	void OnBattleEnd(Scene* map_scene);
+
+	// FIXME: Use C++20 span
+	static void OnMapChange(std::vector<Game_Picture>& pictures);
```
Comment (Ghabry) [inline src/game_picture.h]: these span fixme again

**Comment by fmatthew5876** [comment]
> Question: When ending battle you don't call Take before Append. Can this have any negative side effects? Double free of Drawables e.g.

The `DrawableList` doesn't own the drawables, and since battle is ending the `Scene_Battle` list is about to be destroyed, so we don't need to remove them.

> 
> Problem:
> 
>     1. Create a picture on the map that is not visible in battle
> 
>     2. Start battle
> 
>     3. Call show picture with the same picture ID but this time visible in battle. _Currently the pic is not shown_

Great catch, I put in a small fix for this.


Overall I don't like this model we have with drawable list. It's hard to reason about which list each drawable is on and easy to have dangling pointers.

Also calling `Take()` is O(n) for each picture, making the whole battle start sequence O(map list size * num pictures)  which is obviously sub-optimal.

I have plans to refactor and clean this up somehow, but I haven't yet decided on the best approach. Anyway, that's for a later PR.

File: src/game_picture.h (hunk 1 above)
Comment (fmatthew5876) [reply src/game_picture.h]: Removed

**Comment by fmatthew5876** [comment]
Retested this on windows.

**Comment by Ghabry** [comment]
The drawable list is also a more historical concept.
Initially we did the scene handling like RPG XP and only had one scene without any Push/Pop concept though this was very inflexible because when returning from menus the menu had to know which scene to go to and you had to serialize all scene state.
VX Ace switched to a scene stack, guess they noticed independently from us that there solution was bad :)
With the Push/Pop added we started to encounter problems with other scenes drawing each other so the pushable graphic stack was the solution for it but I agree that drawable transfer between scenes is tricky, that was not part of the design because until RPG2k3E this wasn't really needed :/.

**Comment by fmatthew5876** [comment]
Scene stack is absolutely the way to go.

 I am thinking maybe drawables should be `unique_ptr` owned directly by the list. That would cleanup the confusion and enforce correctness.

**Pull request merged**